### DISCUSSION TITLE
Reading e2e metadata

### DISCUSSION THREAD

**Discussion by BLMeltdown** [marksgraham/OCT-Converter#28] [General]
HI, thanks for this very interesting tool.
Is it possible to read metadata for the exam, especially for e2e exams, please?
Thanks
Best regards
Laurent

**Comment by marksgraham** [comment]
Hi,
What sort of metadata are you interested in?
Currently for e2e, the only metadata extracted is the laterality. It should be possible to extract more - this [uocte wiki](https://bitbucket.org/uocte/uocte/wiki/Heidelberg%20File%20Format) gives an idea of what metadata is currently know about in e2e files

**Comment by amanasj** [comment]
Hello, thank you for the great package.  When trying to extract the laterality info using:

for volume in oct_volumes:
    volume.save('{}.png'.format(volume.laterality))

I get files saved as None.png.  I'm sure I'm doing something obvious wrong here but I'd appreciate a pointer.

Thanks

**Reply by marksgraham** [reply]
Hi,

You can see here 
https://github.com/marksgraham/OCT-Converter/blob/b6fc55075d2e8551440637825489a9c3ba2a0a9e/oct_converter/readers/e2e.py#L152-L161

that if the reader fails to read the 20 bytes the laterality data is usually stored in, the laterality gets set to none. So it might be that your .e2e files are slightly different to the ones this code was built/tested on- do the actual image slices get extracted OK?

**Reply by amanasj** [reply]
Hi, thanks for the reply. The images get extracted perfectly, it's just the laterality that I need to batch process for a very large folder of e2e files for the purposes of training a unet segmentation routine.

**Reply by marksgraham** [reply]
Are you able to share any .e2e files for me to take a look at? If data protection means no - do you know the laterality of any of the images you're extracting? If so perhaps you could print out 'raw' on line 153 and see if it correlates with what you already know for some of the volumes

**Reply by amanasj** [reply]
Hi Mark

Very much appreciate you having a look. I can't share any of the patient
files unfortunately but do have a mouse eye which I am also struggling to
obtain the laterality info for  I don't know which eye this file is for unfortunately and the en-face images for
mice are almost impossible to tell laterality also. I'm hoping perhaps the
meta-data might be illuminating without knowing a priori which it is.

I tried printing raw or f.read(20) in several places but I really struggle
with python (R man here) and don't know how to make print statements within
modules work.


Could I email the e2e file if possible?

Thanks again
Aman

**Reply by Dbrown411** [reply]
Some semi-related and uncaught errors appear to be preventing the laterality data from being properly set. I will be able to address this pretty quickly, and should be updated upon Mark's review.

**Reply by amanasj** [reply]
That's brilliant! thank you very much

**Comment by amanasj** [comment]


**Reply by Dbrown411** [reply]
Hi aman,

i'm happy to try to help out as well. I have .E2Es from our system to which I could compare yours, and have a little time today.

Could you confirm from what device(s) you're obtaining the scans? Same system between mouse and human?

Best,

Dillon

**Reply by amanasj** [reply]
Sorry, I didn't mean to share the e2e publically, can I email it to you?
The device is the same as far as I know (colleague took them). He says it is the same heidelberg system with heyex software

**Reply by Dbrown411** [reply]
Aman,

If willing, feel free to email/share with me at ----. 

(nevermind, got the share access. downloading now and will see if i can figure something out)

Dillon

**Reply by amanasj** [reply]
Thank you Dillon,  really appreciate any help

**Reply by amanasj** [reply]
could I ask, for my learning, how do you print out the metadata?  If I put print(raw) after line 153, how do I actually action this considering the e2e.py is a module within the main OCT-converter program?

**Reply by Dbrown411** [reply]
This is one way:

1. git clone the repo (will create a folder called OCT-Converter)
2. Create a .py file within the OCT-Converter folder (top level, not within oct_converter). from this location, you are able to import things from the folders as if they are python packages, with each folder representing a module.
3. Include "from oct_converter.readers import E2E" in your imports. this should now import the class E2E from your local file \OCT-Converter\oct_convert\readers\e2e.py
4. now, if you modify the e2e.py file, your script will use that modified version.

If this doesn't work, pip uninstalling oct_converter. not sure which package would take priority.

**Comment by amanasj** [comment]
That is fantastic. Thank you so much for your time.
